### PR TITLE
iterative TV implementation

### PR DESCRIPTION
**Pull request by tjlane**
This PR adds support for the iterative TV algorithm described in the pending Fadini et al publication.

### CODE REVIEW

**Comment by tjlane** [comment]
before
before

after
after

**Comment by tjlane** [comment]
@alisiafadini this is ready for review -- anything that is confusing should be questioned, to see if we can make it better. Beyond the implementation, appreciate your thoughts about the public interface(s), i.e. column labels etc.

We will have to add "realistic" tests, such as your cis/trans example, in a future PR. That one might be a good end-to-end test for the scripts.

File: meteor/iterative.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+    derivative_amplitudes: np.ndarray,
+    difference: np.ndarray,
+) -> np.ndarray:
+    """
+    Project `derivative` onto the set of experimentally observed amplitudes,
+
+        Fh' = (D_F' + F) * [|Fh| / |D_F' + F|]
+
+    In English, the output is a complex-valued array that changes the derivative phase to ensure:
+
```
Comment (alisiafadini) [inline meteor/iterative.py:28]: Confusing docstring. Is derivative phase the "true" new state? In difference map code we have called derivative the second measured state

File: meteor/iterative.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    This function contains the algorithm logic.
+
+    Parameters
+    ----------
+    complex_native: np.ndarray
+        The complex native structure factors, usually experimental amplitudes and calculated phases
+
+    initial_complex_derivative : np.ndarray
+        The complex derivative structure factors, usually with experimental amplitudes and esimated
```
Comment (alisiafadini) [inline meteor/iterative.py:71]: estimated (typo). This docstring is more clear than above!

File: meteor/iterative.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+    native_amplitude_column: str = "F",
+    derivative_amplitude_column: str = "Fh",
+    calculated_phase_column: str = "PHIC",
+    output_derivative_phase_column: str = "PHICh",
+    convergence_tolerance: float = 1e-3,
+    max_iterations: int = 100,
+    tv_weights_to_scan: list[float] = [0.001, 0.01, 0.1, 1.0],
+) -> tuple[rs.DataSet, pd.DataFrame]:
+    """
+    Here is a brief psuedocode sketch of the alogrithm. Structure factors F below are complex unless
```
Comment (alisiafadini) [inline meteor/iterative.py:146]: typo

File: meteor/iterative.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+        while not converged:
+            D_rho = FT{ D_F }                       Fourier transform
+            D_rho' = TV{ D_rho }                    TV denoise: apply real space prior
+            D_F' = FT-1{ D_rho' }                   back Fourier transform
+            Fh' = (D_F' + F) * [|Fh| / |D_F' + F|]  Fourier space projection onto experimental set
+            D_F = Fh' - F
+
+    Where the TV lambda parameter is determined using golden section optimization. The algorithm
+    iterates until the changes in DF for each iteration drop below a specified per-amplitude
```
Comment (alisiafadini) [inline meteor/iterative.py:164]: how do we pick the per-amplitude tolerance?

File: meteor/iterative.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        )
+
+        denoised_difference = rs_dataseies_to_complex_array(
+            denoised_map_coefficients[delta_amp.name], denoised_map_coefficients[delta_phase.name]
+        )
+
+        return denoised_difference, tv_metadata
+
+    # convert the native and derivative datasets to complex arrays
+    native = rs_dataseies_to_complex_array(
```
Comment (alisiafadini) [inline meteor/iterative.py:241]: I changed this in the diffmap branch: should fix the typo in this function name

File: meteor/iterative.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+        tv_denoise_function=tv_denoise_closure,
+        convergence_tolerance=convergence_tolerance,
+        max_iterations=max_iterations,
+    )
+    _, derivative_phases = complex_array_to_rs_dataseries(
+        it_tv_complex_derivative, input_dataset.index
+    )
+
+    # combine the determined derivative phases with the input to generate a complete output
+    output_dataset = input_dataset.copy()
```
Comment (alisiafadini) [inline meteor/iterative.py:261]: do we want to copy or just add a column to the og dataset? From our discussions I thought we have converged to adding a column

File: test/unit/conftest.py [hunk 7]
```diff
@@ -90,3 +90,29 @@ def noise_free_map() -> rs.DataSet:
 @pytest.fixture
 def noisy_map() -> rs.DataSet:
     return displaced_single_atom_difference_map_coefficients(noise_sigma=0.03)
+
+
+@pytest.fixture
+def atom_and_noisy_atom() -> rs.DataSet:
```
Comment (alisiafadini) [inline test/unit/conftest.py:96]: what is this? How could I use it in other tests? better name?

File: test/unit/test_iterative.py [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+from skimage.data import binary_blobs
+from skimage.restoration import denoise_tv_chambolle
+
+from meteor import iterative
+from meteor.testing import assert_phases_allclose
+from meteor.tv import TvDenoiseResult
+from meteor.utils import compute_map_from_coefficients
+
+
+def map_norm(map1: gemmi.Ccp4Map, map2: gemmi.Ccp4Map) -> float:
```
Comment (alisiafadini) [inline test/unit/test_iterative.py:16]: I wonder if we should add a test that looks at the negentropy metric for the map before/after itTV. What do you think?

File: meteor/iterative.py [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+            break
+
+    return derivative, pd.DataFrame(metadata)
+
+
+def iterative_tv_phase_retrieval(
+    input_dataset: rs.DataSet,
+    *,
+    native_amplitude_column: str = "F",
+    derivative_amplitude_column: str = "Fh",
```
Comment (alisiafadini) [inline meteor/iterative.py:138]: this seems as good as it can be to me right now. What were your concerns with the input/output?

**Review by alisiafadini — APPROVED** [review]

File: test/unit/conftest.py (hunk 7 above)
Comment (tjlane) [reply test/unit/conftest.py:96]: good point, updated accordingly

File: meteor/iterative.py (hunk 6 above)
Comment (tjlane) [reply meteor/iterative.py:262]: discussed in person, conclusion -- return copy with new column in these situations

File: test/unit/test_iterative.py (hunk 8 above)
Comment (tjlane) [reply test/unit/test_iterative.py:16]: added negentropy check to test below

**Pull request merged**